### ISSUE TITLE
components are loaded multiple times

### ISSUE THREAD

**Issue by grosser** [sparkleformation/sparkle_formation#114]
put a `$stdout.puts 'HELLOOOO'` into a component ... run print ... printed twice ...

@chrisroberts

**Comment by chrisroberts** [comment]
Do you have test files that you can provide? I cannot reproduce this behavior locally. Local structure:

```ruby
# sparkleformation/components/comp.rb
SparkleFormation.component(:comp) do
  $stdout.puts 'TEST'
  fubar true
end
```

```ruby
# sparkleformation/test.rb
SparkleFormation.new(:test).load(:comp).overrides do
  testing true
end
```

And running it:

```
→ be sfn print --file test
TEST
{
  "Fubar": true,
  "Testing": true
}
```

**Comment by grosser** [comment]
was trying to reproduce and ran into some other foobar ...

try this: https://github.com/grosser/sparkle-load

getting ERROR: LoadError: cannot load such file -- idn ... tried on Osx and ubuntu ... same result ...

**Comment by grosser** [comment]
check out the `test` branch for duplication reproduction ...
moving the same code into root just blows everything up

**Comment by grosser** [comment]
moving `something` to `sparkleformation` still has duplication issue

**Comment by grosser** [comment]
renaming helpers to something else -> same duplication

**Comment by chrisroberts** [comment]
The contents of the component file are not a SparkleFormation component. The format of them can be seen here: http://www.sparkleformation.io/docs/sparkle_formation/building-blocks.html#components

The files are loaded into packs internally. This is the origin of the duplicate prints and is not an issue with proper use.

**Comment by grosser** [comment]
can you have a look at that idn error too, that was super weird ... placing
things in root seems to break stuff badly

**Comment by grosser** [comment]
also:
how about blowing up then ... better then silently failing ?

raise 'Component xyz did not declare a component'
-> fast feedback and no weirdness

**Comment by grosser** [comment]
... and the file is STILL loaded twice ... 

update the test branch to be:

```
$stdout.puts 1111

SparkleFormation.component(:helpers) do

end
```

loading twice it not a big deal when the definition just overrides itself, but it still means that all components/files are loaded twice -> waste of time + breaks usecases that rely on them being loaded once


how about using require instead of instance_evaling things ?

**Comment by grosser** [comment]
can you reopen / this is not solved ?

**Comment by luckymike** [comment]
@grosser the double loading is by design, removing them would introduce regressions for existing use cases. 

>The files are loaded into packs internally. This is the origin of the duplicate prints and is not an issue with proper use.

As far as your concern regarding use cases that rely on something being loaded only once, I'm having trouble conceiving of such a use case, even as an example. The only thing I can imagine would be a scenario where the act of loading the template triggered a modification to an external file or resource, and I would consider that well beyond proper use of the SparkleFormation DSL. 

If you have an existing implementation you expect to work that fails as a result of double loading, can you open a new issue with a code example so we can suggest the "proper" SparkleFormation approach to handling it?

**Comment by grosser** [comment]
We are doing shady stuff in our component like monkey-patching things and
loading data.
How is double-loading a good thing, can these packs not share the same
cache ?
afaik it's only the memoize being scoped to the instance instead of being
global ...

**Comment by chrisroberts** [comment]
A few things:

> We are doing shady stuff in our component like monkey-patching things

Don't do this. This should not be done via a component file (or any of the other types). The library has an expectation around what it is loading, which is described within the documentation. Usage outside of that description can render unexpected results and should be considered undefined behavior. If some requirement is driving patching via a component file, then adding a check on the class prior to patching will prevent duplicate patches from being applied. However, monkey patching should not be originating from these files. A better implementation would be to load a remote library (or even a local file) during setup to apply that patching. The callback configuration within sfn is an easy place to automatically load extra libraries.

> and loading data.

Depending on the data, being loaded, this should be done via a registry item, and should be within the register block so it is only evaluated as needed.

> How is double-loading a good thing, can these packs not share the same cache

Double loading happens currently due to template path discovery to provide compat support and behavior for pre-1.0 usage. At some point this will likely be removed, but for the time being it will not. This still circles back to non-supported implementation (not what is described as supported in the documentation) generating undefined behavior. 

> afaik it's only the memoize being scoped to the instance instead of being global

As it should. There should be no assumption that this library is only being run in a single use, one off context.

Hope that helps. Thanks for the feedback.

**Comment by grosser** [comment]
thx for the details!

putting the require inside the .sfn worked (that is what you meant with
callback ?)
but it's unusable since any error results in a undebuggable 'ERROR:
RuntimeError: Failed to load configuration from file
(/Users/mgrosser/Code/zendesk/podzilla/pod/.sfn)'
 ... second case of that bogo swallowing all backtraces :(
 ... with BOGO_DEBUG=1 it works, but I don't want my coworkers to have to
learn that ... so staying with components for now (added 'require
"../lib/helpers"' to avoid the duplicate loading)

registry for small things feels messy since everything is supposed to have
it's own file ... could group a few small items together I guess ... and it
does not cache things which is annoying ... so everything needs it's @foo
||= begin ... end

**Comment by luckymike** [comment]
Here's the callbacks documentation:
http://www.sparkleformation.io/docs/sfn/callbacks.html, you'd do a `callbacks.require` in the .sfn config file.

Registry for small things is the encouraged pattern, and it's very empowering as those small things change. That said, you are not limited to a single registry entry per file, you can do the following:

```
$ cat my_registry_file.rb

SfnRegistry.register(:foo) do
  awesome_cool_registry_things
end

SfnRegistry.register(:bar) do
  other_handy_stuff
end

SfnRegistry.register(:baz) do
  something_we_added_for_carl
end
```

You get the idea.

**Comment by grosser** [comment]
callbacks.require worked out nicely and it did not swallow any errors when
the required file had errors :)
